### ISSUE TITLE
Support alternative image selection methods

### ISSUE THREAD

**Issue by jgriffiths1993** [jenkinsci/openstack-cloud-plugin#45]
We rebuild a lot of our OpenStack images frequently (usually weekly) which causes their ImageIDs to change when they're rebuilt.

This means that someone needs to manually go and change the image ID in Jenkins every time an image has been rebuilt.

This would allow a configurable type of image selector. For example, we currently use a selection of image properties to choose which image to use (e.g. image should have "os=Ubuntu" and "major_version=14.04").

I've already prototyped this (mostly in this commit: https://github.com/jgriffiths1993/openstack-cloud-plugin/commit/67581864b3a7ae0dbdaf84b5b8c93352e53619c9) which adds a selector for choosing an image based on multiple properties.

**Comment by olivergondza** [comment]
In fact, there is image name configured in Jenkins not the id. Can you ensure your rebuilds have the same name as the old ones? This works for me just fine with snapshots.

**Comment by olivergondza** [comment]
Ad more flexible selection mechanism, the old JClouds plugin have it and it was removed from OpenStack plugin as unnecessary clutter. I do not feel strongly if we should bring it back or not.

[1] https://github.com/jenkinsci/jclouds-plugin/blob/master/jclouds-plugin/src/main/resources/jenkins/plugins/jclouds/compute/JCloudsSlaveTemplate/config.jelly#L43-L74

**Comment by jgriffiths1993** [comment]
@olivergondza Jenkins actually stores the imageId, not the image name; at least in the master branch.

### Extract from config.xml ###
```
<jenkins.plugins.openstack.compute.JCloudsSlaveTemplate>
    <name>mySlaveTemplateName</name>
    <imageId>b44fa67d-ax7a-6811-b304-064x7934e7ee</imageId>
    <hardwareId>18dc1433-1ee7-40c7-9ca9-ease0e1cac9e</hardwareId>
    <labelString>myLabelString</labelString>
    <userDataId></userDataId>
    <numExecutors>1</numExecutors>
    <fsRoot>/var/jenkins</fsRoot>
    <installPrivateKey>false</installPrivateKey>
    <overrideRetentionTime>0</overrideRetentionTime>
    <keyPairName></keyPairName>
    <networkId>3a8e2396-3b6b-4cc1-bd7c-88e926981243</networkId>
    <securityGroups>mySecGroup</securityGroups>
    <credentialsId>7nsdd719-7eas-4792-8b07-6494755fbb3a</credentialsId>
    <slaveType>SSH</slaveType>
    <availabilityZone></availabilityZone>
</jenkins.plugins.openstack.compute.JCloudsSlaveTemplate>
```

The way that JClouds does this is actually pretty clunky but the features are ultimately more important than the interface _clunkiness_ . I suspect that we could do this more cleanly than the JClouds menu does though.

**Comment by olivergondza** [comment]
> Jenkins actually stores the imageId, not the image name

Oh, you are right. So would it work for you if it resolve the name instead?

**Comment by olivergondza** [comment]
Anyway, any other criteria (except for id equality) can potentially match multiple images. I guess picking one at random is fine.

**Comment by jgriffiths1993** [comment]
Yeah; I think anything other than ID is workable, picking one at random if multiple images are returned.

**Comment by olivergondza** [comment]
@mavlyutov, I intend to change it from id to name. Any objections?

**Comment by mavlyutov** [comment]
go ahead

**Comment by olivergondza** [comment]
Fixed in #58, will be in 2.1
